service/history/queuev2/queue_state.go

```diff
@@ -55,3 +55,16 @@ func (s *VirtualSliceState) TrySplitByTaskKey(taskKey persistence.HistoryTaskKey
 			Predicate: s.Predicate,
 		}, true
 }
+
+func (s *VirtualSliceState) TrySplitByPredicate(predicate Predicate) (VirtualSliceState, VirtualSliceState, bool) {
+	if predicate.Equals(&universalPredicate{}) || predicate.Equals(&emptyPredicate{}) || predicate.Equals(s.Predicate) {
+		return VirtualSliceState{}, VirtualSliceState{}, false
+	}
+	return VirtualSliceState{
+			Range:     s.Range,
+			Predicate: And(s.Predicate, predicate),
+		}, VirtualSliceState{
+			Range:     s.Range,
+			Predicate: And(s.Predicate, Not(predicate)),
```

Comment:
`And`ing predicates is a confusing concept. Looking at the corresponding tests make the behavior clear but it's still not an "and" operation. It acts as union in some cases and overlap in others. Naming confusion aside, I think we would benefit from adding extra validation (in predicate op tests) to ensure the 2 slices returned here do not overlap. e.g.
```
s1 := and(a, b) 
s2 := and(a, not(b))
expect(empty, intersect(s1,s2))
```

Reply:
Added new test cases to test the behavior